feat: API 공통 응답, 예외 처리 기능 추가

PNoahKR opened the pull request:
## 📝작업 내용

- 공통 응답 코드 생성
응답 코드들을 모아둔 Enum클래스 생성, 정상값과 예외값이 합쳐있어 CustomResponseCode로 지음

- API 공통 응답 처리 기능 생성
응답객체클래스와 응답생성클래스로 분리해 생성

- API 공통 예외 처리 기능 생성
Exception.class와 RuntimeException을 상속받은 CustomException.class 예외처리

### 스크린샷

## 💬리뷰 요구사항

- CustomResponseCode에 추가하거나 삭제해야할 코드가 있다면 편하게 말씀해주시길 바랍니다.

## 🔗참고 링크

## ⏰기한

- 2024-09-10

PNoahKR committed:
feat: API 공통 응답, 예외 처리 기능 추가
- API 공통 응답 처리 기능 생성
- 공통 응답 코드 생성
- API 공통 예외 처리 기능 생성

evan-song2 commented on module-common/src/main/java/com/fashionmall/common/response/CustomResponseCode.java:
```diff
@@ -0,0 +1,28 @@
+
+@Getter
+@RequiredArgsConstructor
+public enum CustomResponseCode {
+
+    //공통 코드
+    SUCCESS(HttpStatus.OK, "성공"),
```
사실상 Error처리를 위한 enum이 되는 것 같은데 SUCCESS를 넣음으로써 CustomResponseCode로 통합적으로 사용하게 하시려는 것 같아보입니다.

실패에 대한 ErrorResponseCode로 안하고 공통enum으로 관리하게끔 개발하신 이유가 궁금하네요

PNoahKR replied:
@evan-song2 200 OK, 201 created 등 API 시나리오에 따라 다양하게 넣을 수 있도록 만들었습니다! 200 OK 통일된 지금은 ErrorResponseCode로 수정 후 exception 패키지로 옮기겠습니다!

PNoahKR replied:
@evan-song2 [a8603a5](https://github.com/team-fashionmall/fashionmall/pull/4/commits/a8603a552c83bf5d788d4595796b94e44c85f508) 수정 사항 반영했습니다.
+ 팩키지를...팩토리라고 잘못명시했습니다..🥲

evan-song2 replied:
@PNoahKR ErrorResponseCode에 OK가 있는게 이상한것 같습니다. 다른곳으로 이동시키는건 어떠실까여

PNoahKR replied:
@evan-song2 [90823ec](https://github.com/team-fashionmall/fashionmall/pull/4/commits/90823ec6b3acb380e6cac4ff3f35e299eec99274) 성공응답값을 ApiResponseUtil로 옮겼습니다.

evan-song2 commented:
@PNoahKR  패키지구조를 아래처럼 가져가시는건 어떠실까요?

```
common
 └── exception
      └── CustomException.java
      └── CustomResponseCode.java (성공도 유지하는 enum이라면 response패키지 하위로?)
      └── handler
           └── GlobalExceptionHandler.java
 └── response
      └── CommonResponse.java
 └── util
      └── ApiResponseUtil.java

```

PNoahKR committed:
refactor: 클래스 명 변경, 팩토리 위치 이동
- CustomResponseCode -> ErrorResponseCode
- ErrorResponseCode 위치를 exception 팩토리로 이동

PNoahKR committed:
refactor: 성공 응답 코드 이동
- ErrorResponseCode -> ApiResponseUtil

evan-song2 reviewed (approved):
고생하셨습니다
은미님이 아직 organization 참여가 안되어있으셔서 
이번 PR은 잠시 기다렸다가 merge 하면 좋을 것 같네요!

minnieming commented:
코드 작성시, 필요한 exception이 있으면 CustomResponseCode에 추가해서 사용하면 될까요?

다른 부분들은 클래스 구조부터 헷갈려서 좀 더 공부해본 뒤, 궁금한거 있으면 여쭤보겠습니다!

evan-song2 commented:
@minnieming 
RestControllerAdvice 설정을 사용해서 항상 동일한 형태로 응답값을 내려주도록 설정하는 부분입니다.
예를들면 통신 성공시 아래와 같이 내려가는 구조겠네요.
```
{
    "code": 200,
    "message": "",
    "data" : {
        "id" : 1,
        "createdAt" : "2020-10-10 23:59:59",
        "updatedAt" : "2020-10-10 23:59:59"
    }
}
```

반대로 `throw new CustomException(ErrorResponseCode.FORBIDDEN);` 이 로직중 발생한다면
RestControllerAdvice에서 오류를 핸들링해서 아래와 같이 응답을 내려줍니다
```
{
    "code": 403,
    "message": "권한이 없습니다.",
    "data" : null
}
```
